stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -345,12 +345,11 @@ julia> kron(A, B)
 function kron(a::AbstractMatrix{T}, b::AbstractMatrix{S}) where {T,S}
     require_one_based_indexing(a, b)
     R = Matrix{promote_op(*,T,S)}(undef, size(a,1)*size(b,1), size(a,2)*size(b,2))
-    m = 1
-    for j = 1:size(a,2), l = 1:size(b,2), i = 1:size(a,1)
+    m = 0
+    @inbounds for j = 1:size(a,2), l = 1:size(b,2), i = 1:size(a,1)
         aij = a[i,j]
         for k = 1:size(b,1)
-            R[m] = aij*b[k,l]
-            m += 1
+            R[m += 1] = aij*b[k,l]
```

Comment:
This isn't really a performance optimization; it lowers to the same thing:
```
julia> f(A, x) = A[x += 1]
f (generic function with 1 method)

julia> @code_lowered f([1,2,3], 1)
CodeInfo(
1 1 ─      x@_4 = x@_3                                                                                                                                      │
  │   %2 = x@_4 + 1                                                                                                                                         │
  │        x@_4 = %2                                                                                                                                        │
  │   %4 = (Base.getindex)(A, %2)                                                                                                                           │
  └──      return %4                                                                                                                                        │
)

julia> g(A, x) = (A[x]; x += 1)
g (generic function with 1 method)

julia> @code_lowered g([1,2,3], 1)
CodeInfo(
1 1 ─      x@_4 = x@_3                                                                                                                                      │
  │        (Base.getindex)(A, x@_4)                                                                                                                         │
  │   %3 = x@_4 + 1                                                                                                                                         │
  │        x@_4 = %3                                                                                                                                        │
  └──      return %3                                                                                                                                        │
)
```
The only difference is the order, which shouldn't matter here.

As Andy alluded to, I would be willing to bet that all of the performance difference here is due to `@inbounds`.

Reply:
Well it is strange, maybe some artifact due to using `Revise`? This is my session (no `@inbounds`):

```
1.2.0-DEV.183> @benchmark kron($a,$a)
BenchmarkTools.Trial: 
  memory estimate:  1.22 MiB
  allocs estimate:  2
  --------------
  minimum time:     126.667 μs (0.00% GC)
  median time:      128.983 μs (0.00% GC)
  mean time:        156.719 μs (6.61% GC)
  maximum time:     26.287 ms (98.50% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.183> @benchmark kron($a,$a)
BenchmarkTools.Trial: 
  memory estimate:  1.22 MiB
  allocs estimate:  2
  --------------
  minimum time:     124.069 μs (0.00% GC)
  median time:      129.856 μs (0.00% GC)
  mean time:        226.579 μs (9.96% GC)
  maximum time:     26.276 ms (98.48% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.183> Revise.track(Base)

1.2.0-DEV.183> @benchmark kron($a,$a)
BenchmarkTools.Trial: 
  memory estimate:  1.22 MiB
  allocs estimate:  2
  --------------
  minimum time:     108.931 μs (0.00% GC)
  median time:      110.993 μs (0.00% GC)
  mean time:        143.684 μs (11.25% GC)
  maximum time:     38.443 ms (99.04% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.183> @benchmark kron($a,$a)
BenchmarkTools.Trial: 
  memory estimate:  1.22 MiB
  allocs estimate:  2
  --------------
  minimum time:     109.051 μs (0.00% GC)
  median time:      110.971 μs (0.00% GC)
  mean time:        144.379 μs (11.25% GC)
  maximum time:     35.658 ms (98.84% GC)
  --------------
  samples:          10000
  evals/sample:     1

```

Reply:
Could be. Try defining both versions locally in the REPL and benchmarking.

Reply:
Huh, yeah, I can.